### PR TITLE
Add infini-scroll

### PR DESCRIPTION
**Pull request by mruwnik**
image
I added a loader type thingy, but left the footer for now

### CODE REVIEW

File: app/hooks/useQuestionStateInUrl.ts [hunk 1]
```diff
@@ -1,22 +1,25 @@
...
 
-const getStateEntries = (state: string): [string, QuestionState][] =>
+type PageId = Question['pageid']
+
+const getStateEntries = (state: string): [PageId, QuestionState][] =>
   Array.from(state.matchAll(/([^-_r]+)([-_r]*)/g) ?? []).map((groups) => [
-    groups[1],
-    (groups[2] || '_') as QuestionState,
+    groups[1], // question id
+    (groups[2] as QuestionState) || QuestionState.OPEN,
```
Comment (mruwnik) [inline app/hooks/useQuestionStateInUrl.ts:12]: I was finding it hard to remember what char meant what, so I went and made an enum for it

File: app/hooks/useQuestionStateInUrl.ts [hunk 2]
```diff
@@ -83,17 +86,111 @@ export default function useQuestionStateInUrl(minLogo: boolean, initialQuestions
...
+  )
+
+  /*
+   * Add the given `questions` to the global questions map. This will not update the URL
+   */
+  const mergeNewQuestions = useCallback((questions: Question[]) => {
+    setQuestionMap((currentMap) =>
+      questions.reduce((map, question) => updateQuestionMap(question, map), new Map(currentMap))
+    )
+  }, [])
```
Comment (mruwnik) [inline app/hooks/useQuestionStateInUrl.ts:162]: the previous functions were simply extracted from `toggleQuestion` so they could be used elsewhere

File: app/hooks/useQuestionStateInUrl.ts [hunk 3]
```diff
@@ -83,17 +86,111 @@ export default function useQuestionStateInUrl(minLogo: boolean, initialQuestions
...
+
+  /*
+   * Add the given `newQuestions` to the collection of questions. This will also update the URL
+   */
+  const addQuestions = useCallback(
```
Comment (mruwnik) [inline app/hooks/useQuestionStateInUrl.ts:167]: this function is new and is the main real change in this file

**Comment by Aprillion** [comment]
> but left the footer for now

as I wrote in #77 I don't agree with keeping both the infinite scrolling and the footer.

File: app/server-utils/stampy.ts [hunk 4]
```diff
@@ -2,7 +2,12 @@ import {withCache} from '~/server-utils/kv-cache'
 import MarkdownIt from 'markdown-it'
 import MarkdownItFootnote from 'markdown-it-footnote'
 
-export type QuestionState = '_' | '-' | 'r'
+// export type QuestionState = '_' | '-' | 'r'
```
Comment (Aprillion) [inline app/server-utils/stampy.ts:5]: if we don't need to keep commented-out code for some special reason, please delete.

File: app/server-utils/stampy.ts [hunk 5]
```diff
@@ -2,7 +2,12 @@ import {withCache} from '~/server-utils/kv-cache'
...
+export enum QuestionState {
+  OPEN = '_',
+  COLLAPSED = '-',
+  SUBQUESTION = 'r',
```
Comment (Aprillion) [inline app/server-utils/stampy.ts:9]: `r` was a short for RELATED question

File: app/root.css [hunk 6]
```diff
@@ -26,6 +26,25 @@
   /* dark theme at the end of file to overwrite all previous rules */
 }
 
+.loader {
+  border: 4px solid #f3f3f3; /* Light grey */
+  border-top: 4px solid #3498db; /* Blue */
```
Comment (Aprillion) [inline app/root.css:31]: do both colors work well in both light and dark theme?

File: app/components/infiniteScroll.tsx [hunk 7]
```diff
@@ -0,0 +1,60 @@
+import {ReactNode, useState, useRef, useCallback} from 'react'
+
+type Props = {
+  children: ReactNode[]
+  perPage: number
+  fetchMore: (page: number, perPage: number) => any[]
+  loadingRenderer: (loading: boolean) => ReactNode
+  [k: string]: any
```
Comment (Aprillion) [inline app/components/infiniteScroll.tsx:8]: is there some non-any type we could use, please? e.g. `JSX.IntrinsicElements['div']` or `React.HTMLAttributes<HTMLDivElement>`?

File: app/server-utils/stampy.ts (hunk 4 above)
Comment (mruwnik) [reply app/server-utils/stampy.ts:5]: whoops - good catch

File: app/components/infiniteScroll.tsx (hunk 7 above)
Comment (mruwnik) [reply app/components/infiniteScroll.tsx:8]: nice!

File: app/root.css (hunk 6 above)
Comment (mruwnik) [reply app/root.css:31]: image
Depends on what you want :) To be honest, this was pretty much the first css loader that I found, so any corrections/enhancements are welcome

File: app/server-utils/stampy.ts (hunk 5 above)
Comment (mruwnik) [reply app/server-utils/stampy.ts]: updated

File: app/root.css (hunk 6 above)
Comment (Aprillion) [reply app/root.css:31]: yeah, visibe enough for now, can be improved later if someone will find a better color

**Review by Aprillion — APPROVED** [review]
Should be good-enough to merge with the floating join discord icon, `Dashboard` and `Feedback` are less critical to figure out .. as long as @plexish or someone is keeping track what needs to move where.

Leaving copyright in footer despite infinite scrolling should be fine, no user reads it and developers can check inside code.

**Comment by plexish** [comment]
Feedback got two responses total despite me promoting it, and Dashboard is.. not essential, it's mostly useful for reviewers who should be able to find their way there anyway as they're meant to be familiar with the system and it's only one extra click from Get Involved. This can go in now that there's a Discord link, thanks for making it!

#77 mentioned using wiki data (sorry if you followed that dated spec!), but as the wiki's going away using coda is the way to go. Using the ordering on Initial Answers (allowing non-Integer values, so it's easier to re-order), and using the variable formula in that to determine how many to display initially.

**Comment by mruwnik** [comment]
Currently it loads all the questions from the `initial questions` page, and if you then scroll down, will fetch another 10 questions from the `On-site API` table, skipping any questions that have already been displayed. This can result in slightly fewer than 10 questions per load, but I'm guessing that's fine :D

**Comment by plexish** [comment]
Okay, that seems reasonable. I can make `On-site API` sorted by something sensible later.

**Review by Aprillion — APPROVED** [review]

**Pull request merged**